source/geod24/concurrency.d

```diff
@@ -1012,56 +1178,154 @@ private:
 
         Start the scheduling loop
 
-        Param:
-            pos = m_fibers index of the Fiber to execute first
-
     ***********************************************************************/
 
-    void dispatch(size_t pos = 0)
+    void dispatch ()
     {
-        import std.algorithm.mutation : remove;
+        thisScheduler(this);
+        scope (exit) thisScheduler(null);
+
+        MonoTime earliest_timeout;
 
-        assert(pos < m_fibers.length);
-        while (m_fibers.length > 0)
+        while (!this.readyq.empty() || !this.wait_list.empty())
         {
-            // Is Fiber waiting on a FiberBinarySemaphore?
-            if (auto sem = m_fibers[pos].sem)
+            while (!readyq.empty())
             {
-                // Is condition met?
-                // TRUE: Clear the sem and schedule the fiber
-                // FALSE: Skip it
-                if (sem.shouldBlock())
+                InfoFiber cur_fiber = this.readyq.front();
+                this.readyq.removeFront();
+
+                assert(cur_fiber.state != Fiber.State.TERM);
+
+                auto t = cur_fiber.call(Fiber.Rethrow.no);
+
+                // Fibers that block on a FiberBlocker throw an
+                // exception for scheduler to catch
+                if (t is this.blocked_ex)
                 {
-                    if (pos++ >= m_fibers.length - 1)
-                        pos = 0;
+                    auto cur_timeout = cur_fiber.blocker.getTimeout();
+
+                    // Keep track of the earliest timeout in the system
+                    if (cur_timeout != MonoTime.init
+                            && (earliest_timeout == MonoTime.init || cur_timeout < earliest_timeout))
+                    {
+                        earliest_timeout = cur_timeout;
+                    }
+
+                    this.wait_list.insert(cur_fiber);
                     continue;
                 }
-                else
+                else if (t)
+                {
+                    // We are exiting the dispatch loop prematurely, all resources
+                    // held by Fibers should be released.
+                    this.releaseResources(cur_fiber);
+                    throw t;
+                }
+
+                if (cur_fiber.state != Fiber.State.TERM)
                 {
-                    m_fibers[pos].sem = null;
+                    this.readyq.insert(cur_fiber);
                 }
+
+                // See if there are Fibers to be woken up if we reach a timeout
+                // or the scheduler semaphore was notified
+                if (MonoTime.currTime >= earliest_timeout || this.sem.tryWait())
+                    earliest_timeout = wakeFibers();
             }
 
-            auto t = m_fibers[pos].call(Fiber.Rethrow.no);
-            if (t !is null)
+            if (!this.wait_list.empty())
             {
-                throw t;
+                Duration time_to_timeout = earliest_timeout - MonoTime.currTime;
```

Comment:
Is it guaranteed that `earliest_timeout` will not equal `Duration.init` here? Actually I'm not even sure what happens when we subtract against that.

Reply:
Yes, `earliest_timeout` is always > 0